python/ray/monitor.py

```diff
@@ -468,6 +475,77 @@ def driver_removed_handler(self, unused_channel, data):
 
         self._clean_up_entries_for_driver(driver_id)
 
+    def _xray_clean_up_entries_for_driver(self, driver_id):
+        """Remove this driver's object/task entries from redis.
+
+        Removes control-state entries of all tasks and task return
+        objects belonging to the driver.
+        """
+        redis = self.state.redis_client
+
+        task_table_infos = {}  # task id -> TaskInfo
+        for key in redis.scan_iter(match=XRAY_TASK_TABLE_PREFIX + b"*"):
+            entry = redis.get(key)
+            task = Task.GetRootAsTask(entry, 0)
+            task_info = TaskInfo.GetRootAsTaskInfo(task.TaskSpecification(), 0)
+            if driver_id != task_info.DriverId():
+                # Ignore tasks that aren't from this driver.
+                continue
+            task_table_infos[task_info.TaskId()] = task_info
+        driver_task_ids = task_table_infos.keys()
+
+        # Get the list of objects returned by driver tasks.
+        driver_object_ids = []
+        for task_info in task_table_infos.values():
+            driver_object_ids.extend([
+                task_info.Returns(i) for i in range(task_info.ReturnsLength())
+            ])
+
+        # TODO: Remove only objects in GCS that originated from this driver.
```

Comment:
Are you saying we don't have enough information in the GCS to flush objects created by ray.put? From the object ID, we can determine the task ID, and from that we can determine the Driver ID, right? So I think we should be able to flush them already.

Reply:
Also, let's make sure that we're testing cleanup for `ray.put` objects.

Reply:
The object ID does not contain the task ID.

Reply:
What I mean is that the object ID itself tells you the task ID, right? https://github.com/ray-project/ray/blob/aa423318442c18e3e73b6125a5c5237258d60ff0/src/ray/id.cc#L166-L174

Reply:
Excellent point. I was not aware of this method.